### PR TITLE
Update perf testing documentation

### PR DESCRIPTION
**Pull request by paulkagiri**

### CODE REVIEW

File: uiautomationutilities/perf_tool/readme.md [hunk 1]
```diff
@@ -8,148 +8,106 @@
...
-	//Setting up scenario code with a 3 digit code. For example existing codes are 100 and 200, choose a new one.
-	CodeMarkerManager.getInstance().setPrefixScenarioCode("100");
+- If a new scenario has to be created, write a new test case in MSAL native codebase, see existing test case [TestCasePerf.java](https://github.com/AzureAD/microsoft-authentication-library-for-android/blob/dev/msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/perf/TestCasePerf.java) for non-brokered auth and [TestCasePerfBrokered.java](https://github.com/AzureAD/microsoft-authentication-library-for-android/blob/dev/msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/perf/TestCasePerfBrokered.java) for brokered auth.
+- In the new test case, before performing the particular scenario, add following code snippet
+	- Enable the Code marker: `CodeMarkerManager.getInstance().setEnableCodeMarker(true);`
+	- Setting up scenario code with a 3 digit code. For example some of the existing codes are 100 and 200, choose a new one and set it up as: `CodeMarkerManager.getInstance().setPrefixScenarioCode("100");`
```
Comment (iambmelt) [inline uiautomationutilities/perf_tool/readme.md:15]: Are these codes defined in a unified location? Where is that?

File: uiautomationutilities/perf_tool/readme.md [hunk 2]
```diff
@@ -8,148 +8,106 @@
 
 **Pipeline information:** https://dev.azure.com/IdentityDivision/IDDP/_build?definitionId=1254
```
Comment (iambmelt) [inline uiautomationutilities/perf_tool/readme.md:9]: Is this going to live in `IDDP` for the foreseeable future? Or will this pipeline eventually move over to `ENGINEERING`?

File: uiautomationutilities/perf_tool/readme.md [hunk 3]
```diff
@@ -8,148 +8,106 @@
...
+- In the new test case, before performing the particular scenario, add following code snippet
+	- Enable the Code marker: `CodeMarkerManager.getInstance().setEnableCodeMarker(true);`
+	- Setting up scenario code with a 3 digit code. For example some of the existing codes are 100 and 200, choose a new one and set it up as: `CodeMarkerManager.getInstance().setPrefixScenarioCode("100");`
 	- If we are having more than one iterations of same scenario, i.e. running same code multiple times, 
-		- make sure to clear the previous run's markers in the start of the iteration as follows:
-		  "CodeMarkerManager.getInstance().clearMarkers();"
+		- Make sure to clear the previous run's markers in the start of the iteration as follows: `CodeMarkerManager.getInstance().clearMarkers();`
 		- Also write the content to the file using fileAppender.
```
Comment (iambmelt) [inline uiautomationutilities/perf_tool/readme.md:18]: Link to source for `FileAppender`?

File: uiautomationutilities/perf_tool/readme.md [hunk 4]
```diff
@@ -8,148 +8,106 @@
```
Comment (iambmelt) [inline uiautomationutilities/perf_tool/readme.md:8]: Are there written step-by-step instructions for setting up the C# perf tool project? Is that forthcoming?

File: uiautomationutilities/perf_tool/readme.md [hunk 5]
```diff
@@ -8,148 +8,106 @@
...
 ### Steps to deploy latest version of C# tool to pipeline:
-- 1. Take a checkout of C# tool repo (Details above)
-- 2. Open the Location of the perf tool (Location details given above) in Visual studio.
-- 3. Build Solution by clicking on the Build -> Build Solution. Make sure Release build is chosen for deployment.
-- 4. Bundle following files into one zip file named "PerfRunnables.zip":
+- Take a checkout of C# tool repo (Details above)
+- Open the Location of the perf tool (Location details given above) in Visual studio.
+- Build Solution by clicking on the Build -> Build Solution. Make sure Release build is chosen for deployment.
+- Bundle following files into one zip file named "PerfRunnables.zip":
 	- a. Files "PerfIdentity.exe", "PerfIdentity.runtimeconfig.json", "PerfIdentity.dll" and "PerfIdentity.deps.json" from location  "bin\Release\netcoreapp3.1" and file "PerfDataConfiguration.xml" from root location of the project
```
Comment (iambmelt) [inline uiautomationutilities/perf_tool/readme.md:41]:
```suggestion
	- a. Files `PerfIdentity.exe`, `PerfIdentity.runtimeconfig.json`, `PerfIdentity.dll` and `PerfIdentity.deps.json` from location  `bin\Release\netcoreapp3.1` and file `PerfDataConfiguration.xml` from root location of the project
```

File: uiautomationutilities/perf_tool/readme.md [hunk 6]
```diff
@@ -8,148 +8,106 @@
...
-- 6. Click on variables.
-- 7. Change the value of the variable baseBuildID with the value we have noted in the steps above.
-- 8. Save the pipeline definition.
+- Go to the pipeline page where we can see the previous runs.
+- From the list of the runs, Click the particular run which we want to choose as the baseline.
+- From the URL, find the numberic value of the buildId and note it. For example buildId will be 708340 for following URL: https://dev.azure.com/IdentityDivision/IDDP/_build/results?buildId=708340&view=results
+- Go to the run page of the pipeline again.
+- Go to the definition of the pipeline by clicking on the "Edit" button on the top right.
+- Click on variables.
+- Change the value of the variable baseBuildID with the value we have noted in the steps above.
```
Comment (iambmelt) [inline uiautomationutilities/perf_tool/readme.md:54]: Suggest clarifying which value `baseBuildID` refers to.... is it the `scenario` code?

File: uiautomationutilities/perf_tool/readme.md [hunk 7]
```diff
@@ -8,148 +8,106 @@
...
-- 6. Extract and put the raw basefiles in your local base file folder of your machine.
-- 7. On local windows machine go to your target codemarker files location and download the codemarker files named as "PerfData*.txt" from the android device from location "/sdcard" : (Note: make sure that the test case has completed before running the download command)
-- 8. Open the Perf C# tool in visual studio by opening the solution file "PerfIdentity.sln"
-- 9. Build the project by clicking on Build -> Build Solution
--10. Copy PerfDataConfiguration.xml into the build folder e.g. "bin\Release\netcoreapp3.1"
--11. Run the application. Arguments definition can be seen in command.cs file.
+- Checkout msal native code recursively. 
+- Open the msal native code in Android studio and point the dependency builds to local variants.
+- Run one of the Test case having the codemarkers enabled. e.g. "TestCasePerf.java", "TestCasePerfBrokerHost.java" on an Android device.
```
Comment (iambmelt) [inline uiautomationutilities/perf_tool/readme.md:82]:
```suggestion
- Run one of the Test case having the codemarkers enabled. e.g. `TestCasePerf.java`, `TestCasePerfBrokerHost.java` on an Android device.
```

File: uiautomationutilities/perf_tool/readme.md [hunk 8]
```diff
@@ -8,148 +8,106 @@
...
+- On local windows machine create directories to keep your base and target codemarker files. e.g. `C:\testdata\basefiles` and `C:\testdata\targetfiles`
+- Download base files from any of the build run by going to the artifacts, clicking on Perf and then clicking on basefiles. For example go to "https://dev.azure.com/IdentityDivision/IDDP/_build/results?buildId=705958&view=results", click on artifact button "2 published; 1 consumed", click on Perf and then download basefiles. 
+- Extract and put the raw basefiles in your local base file folder of your machine.
+- On local windows machine go to your target codemarker files location and download the codemarker files named as "PerfData*.txt" from the android device from location "/sdcard" : (Note: make sure that the test case has completed before running the download command)
+- Open the Perf C# tool in visual studio by opening the solution file `PerfIdentity.sln`
+- Build the project by clicking on Build -> Build Solution
+- Copy PerfDataConfiguration.xml into the build folder e.g. `bin\Release\netcoreapp3.1`
```
Comment (iambmelt) [inline uiautomationutilities/perf_tool/readme.md:89]:
```suggestion
- Copy `PerfDataConfiguration.xml` into the build folder e.g. `bin\Release\netcoreapp3.1`
```

File: uiautomationutilities/perf_tool/readme.md [hunk 9]
```diff
@@ -8,148 +8,106 @@
...
+- Run the application. Arguments definition can be seen in command.cs file.
```
Comment (iambmelt) [inline uiautomationutilities/perf_tool/readme.md:90]:
```suggestion
- Run the application. Arguments definition can be seen in `command.cs` file.
```

File: uiautomationutilities/perf_tool/readme.md (hunk 4 above)
Comment (iambmelt) [reply uiautomationutilities/perf_tool/readme.md:8]: Oh, I see - it's mentioned later in this file:
https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/e24be64132ffa78369518e1e5c8a0cf466194c18/uiautomationutilities/perf_tool/readme.md#steps-to-run-on-local-machine

Maybe add an index to this document and link to various sections from the top?

Comment (iambmelt) [inline uiautomationutilities/perf_tool/readme.md:8]: There is some terminology used in this document which I think could use a little definition for newcomers/maintainers. Specifically, concepts like `baseFiles` and `targetFiles` are mentioned. As I understand it, a `baseFile` is a "first run" against which subsequent runs (targets?) are compared... can we define these terms explicitly within this document?

File: uiautomationutilities/perf_tool/readme.md [hunk 10]
```diff
@@ -8,148 +8,106 @@
...
-- 5. Device model to be written in the final Email report. Example value: "Pixel2"
-- 6. Device OS to be written in the final Email report. Example value: "API28"
-- 7. App name to be written in the Email report. Example value: "MSALTestApp"
-- 8. Email ID of the sender's account. Example value: "idlab1@msidlab4.onmicrosoft.com"
-- 9. Password of the sender's account.
--10. Email To list separated by comma
+- Directory where PerfData base files are present. Example value: `C:\testdata\basefiles`
+- Directory where PerfData target files are present. Example value: `C:\testdata\targetfiles`
+- Build id of the base task.
+- Build id which should be written in the final Email report and used for going to artifact url. Example value: `1234`
```
Comment (iambmelt) [inline uiautomationutilities/perf_tool/readme.md:107]: Do we have it documented somewhere how the email / reporting system is configured?

File: uiautomationutilities/perf_tool/readme.md (hunk 2 above)
Comment (paulkagiri) [reply uiautomationutilities/perf_tool/readme.md:9]: Am not sure, what are the ramifications of remaining in `IDDP`

File: uiautomationutilities/perf_tool/readme.md (hunk 6 above)
Comment (paulkagiri) [reply uiautomationutilities/perf_tool/readme.md]: `baseBuildID ` is now part of the definitions at the start of the doc

File: uiautomationutilities/perf_tool/readme.md (hunk 4 above)
Comment (paulkagiri) [reply uiautomationutilities/perf_tool/readme.md:10]: Done

File: uiautomationutilities/perf_tool/readme.md (hunk 3 above)
Comment (paulkagiri) [reply uiautomationutilities/perf_tool/readme.md]: Added

File: uiautomationutilities/perf_tool/readme.md (hunk 4 above)
Comment (paulkagiri) [reply uiautomationutilities/perf_tool/readme.md:10]: Done

File: uiautomationutilities/perf_tool/readme.md (hunk 10 above)
Comment (paulkagiri) [reply uiautomationutilities/perf_tool/readme.md:148]: Now we do

File: uiautomationutilities/perf_tool/readme.md [hunk 11]
```diff
@@ -1,155 +1,157 @@
...
+
+**Basefiles:** These are the code marker files during the run of the base build and help us compare the current run against the base run
+
+**Targetfiles:** These are the code marker files generated during the current run
+
+
+### Steps to add a new marker and see it in the report:
+- If a new scenario has to be created, come up with a scenario code comprised of 3 digits and add it to the codemarker definitions in file [PerfConstants.java](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/dev/common/src/main/java/com/microsoft/identity/common/PerfConstants.java), for example some of the existing scenario codes are 100 and 200.
+- In the common codebase, and add codemarkers (e.g. start of the activity to be measured and the end of the activity to be measured) at desired places. See example in [CommandDispatcher.java](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/dev/common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java#L197). Example events lines are given as follows:
+	- `CodeMarkerManager.getInstance().markCode(CodeMarkersConstants.ACQUIRE_TOKEN_SILENT_START);`
```
Comment (iambmelt) [inline uiautomationutilities/perf_tool/readme.md:53]: Is there a convention to follow when adding new `CodeMarkerConstants`? Are are the strings/ints "random"?

File: uiautomationutilities/perf_tool/readme.md [hunk 12]
```diff
@@ -1,155 +1,157 @@
...
+		- Make sure to clear the previous run's markers in the start of the iteration as follows: `CodeMarkerManager.getInstance().clearMarkers();`
+		- Also write the content to the file using [FileAppender](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/dev/uiautomationutilities/src/main/java/com/microsoft/identity/client/ui/automation/logging/appender/FileAppender.java) such as [here](https://github.com/AzureAD/microsoft-authentication-library-for-android/blob/dev/msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/perf/TestCasePerf.java#L119-L124)
+	- Also add followings after performing the particular scenario:
+		- Clear the codemarkers: `CodeMarkerManager.getInstance().clearMarkers();`
+		- Disable the codemarkers: `CodeMarkerManager.getInstance().setEnableCodeMarker(false);`
+- In the C# tool code, open the file [PerfDataConfiguration.xml](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/dev/uiautomationutilities/perf_tool/PerfDataConfiguration.xml) and perform following actions:
```
Comment (iambmelt) [inline uiautomationutilities/perf_tool/readme.md:65]: Can this move 'up' in the instructions? We are sort of switching between actions that need to happen inside of `common` and actions that need to happen inside of `msal`.

For instance this step....
>Write a new test case that targets the flow where code markers have been added, see existing test case....

Is actually going to happen in the `msal` project... if I am preparing a PR, I may want to author the changes in common first and then submit a separate PR to pick up my work in MSAL....

File: uiautomationutilities/perf_tool/readme.md [hunk 13]
```diff
@@ -1,155 +1,157 @@
...
+	- In tag `MeasurementsConfigurations`, add a new `MeasurementsConfiguration` with giving the startMarker and endMarker as given in the step 2.
```
Comment (iambmelt) [inline uiautomationutilities/perf_tool/readme.md:66]: How do I know which `Id` to use? Just make one up? What does `startskip` and `endskip` do?

File: uiautomationutilities/perf_tool/readme.md [hunk 14]
```diff
@@ -1,155 +1,157 @@
...
+	- `CodeMarkerManager.getInstance().markCode(CodeMarkersConstants.ACQUIRE_TOKEN_SILENT_START);`
+	- `CodeMarkerManager.getInstance().markCode(CodeMarkersConstants.ACQUIRE_TOKEN_SILENT_FUTURE_OBJECT_CREATION_END);`
+- Write a new test case that targets the flow where code markers have been added, see existing test case [TestCasePerf.java](https://github.com/AzureAD/microsoft-authentication-library-for-android/blob/dev/msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/perf/TestCasePerf.java) for non-brokered auth and [TestCasePerfBrokered.java](https://github.com/AzureAD/microsoft-authentication-library-for-android/blob/dev/msalautomationapp/src/androidTest/java/com/microsoft/identity/client/msal/automationapp/testpass/perf/TestCasePerfBrokered.java) for brokered auth.
+- In the new test case, before performing the particular scenario, add following code snippet
+	- Enable the Code marker: `CodeMarkerManager.getInstance().setEnableCodeMarker(true);`
+	- Setup the scenario code defined in step #1: `CodeMarkerManager.getInstance().setPrefixScenarioCode("100");`
```
Comment (iambmelt) [inline uiautomationutilities/perf_tool/readme.md:58]: Steps are not numbered when this doc is rendered... referring to things like "step 1" is kind of unclear

File: uiautomationutilities/perf_tool/readme.md [hunk 15]
```diff
@@ -1,155 +1,157 @@
...
+	- Also add followings after performing the particular scenario:
+		- Clear the codemarkers: `CodeMarkerManager.getInstance().clearMarkers();`
+		- Disable the codemarkers: `CodeMarkerManager.getInstance().setEnableCodeMarker(false);`
+- In the C# tool code, open the file [PerfDataConfiguration.xml](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/dev/uiautomationutilities/perf_tool/PerfDataConfiguration.xml) and perform following actions:
+	- In tag `MeasurementsConfigurations`, add a new `MeasurementsConfiguration` with giving the startMarker and endMarker as given in the step 2. 
+	- In tag, `Scenarios`, add a new scenario with the Measurement ID of the MesurementConfiguration just added in previous sub-step.
+- Push the common code and msal native code to dev branch by raising PRs.
+- Push the C# changes of its repo (i.e. common) into dev branch by raising a PR.
+- Deploy C# tool into the pipeline. (Pipeline information given at the start of this doc). We can deploy C# tool even before merging of PR if need be. As this deployment is done after building at the local machine.
```
Comment (iambmelt) [inline uiautomationutilities/perf_tool/readme.md:70]: So... adding new scenarios means the tool needs to be rebuilt and redeployed to the pipeline? Why? There's a plan to remove this step in the future?

File: uiautomationutilities/perf_tool/readme.md [hunk 16]
```diff
@@ -1,155 +1,157 @@
...
+[**Scenario:**](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/dev/uiautomationutilities/perf_tool/PerfDataConfiguration.xml#L18-L25) An E2E scenario that we would be interested in measuring. E.g. Silent Token refresh on a 4G network when requesting a token for graph. A scenario can be associated with one or more measurements.
+
+**Base build:** This is a build against which we compare performance of future builds. It would typically be the last release. It is configured on the Perf pipeline using the `baseBuildID` variable
+
+**Basefiles:** These are the code marker files during the run of the base build and help us compare the current run against the base run
+
+**Targetfiles:** These are the code marker files generated during the current run
+
+
+### Steps to add a new marker and see it in the report:
```
Comment (iambmelt) [inline uiautomationutilities/perf_tool/readme.md:50]: I think this section could benefit from tighter formatting... example below

## Steps to Add a New Marker

### (Android Common) Declare Constants, Update PerfTool XML, and Update Source as Necessary
1. Lorem ipsum dolor sit amet, consectetur adipiscing elit
2. Sed do eiusmod tempor incididunt ut labore et dolore magna aliqua
3. etc etc...

### (MSAL/Broker) Define Test Case[s] and Execute
1. Lorem ipsum dolor sit amet, consectetur adipiscing elit
2. Sed do eiusmod tempor incididunt ut labore et dolore magna aliqua
3. etc etc...

File: uiautomationutilities/perf_tool/readme.md (hunk 15 above)
Comment (iambmelt) [inline uiautomationutilities/perf_tool/readme.md:70]: This seems out of order. Shouldn't I have deployed the updated tool _before_ authoring my changes in MSAL?

File: uiautomationutilities/perf_tool/readme.md [hunk 17]
```diff
@@ -1,155 +1,157 @@
...
 ### Steps to deploy latest version of C# tool to pipeline:
-- 1. Take a checkout of C# tool repo (Details above)
-- 2. Open the Location of the perf tool (Location details given above) in Visual studio.
-- 3. Build Solution by clicking on the Build -> Build Solution. Make sure Release build is chosen for deployment.
-- 4. Bundle following files into one zip file named "PerfRunnables.zip":
-	- a. Files "PerfIdentity.exe", "PerfIdentity.runtimeconfig.json", "PerfIdentity.dll" and "PerfIdentity.deps.json" from location  "bin\Release\netcoreapp3.1" and file "PerfDataConfiguration.xml" from root location of the project
-- 5. Go to the pipeline and then from the left side options, Open Library page and upload the file "PerfRunnables.zip" after deleting previous one (if exists)
-- 6. Go to the pipeline definition and make change in the task "Download Perf runnable package" by selecting the file name which was uploaded in the "secure file" drop down.
-- 7. Save the build definition. 
+- Install a recent version of [Visual Studio](https://visualstudio.microsoft.com/)
```
Comment (iambmelt) [inline uiautomationutilities/perf_tool/readme.md:74]: Probably should call out that Windows is requisite for this. Or is it not?

File: uiautomationutilities/perf_tool/readme.md (hunk 15 above)
Comment (paulkagiri) [reply uiautomationutilities/perf_tool/readme.md]: The PerfDataConfiguration.xml is the only file that needs to be updated - will look for a better way of updating the file without bundling the whole report tool

File: uiautomationutilities/perf_tool/readme.md (hunk 14 above)
Comment (paulkagiri) [reply uiautomationutilities/perf_tool/readme.md]: This has been addressed

File: uiautomationutilities/perf_tool/readme.md (hunk 11 above)
Comment (paulkagiri) [reply uiautomationutilities/perf_tool/readme.md:55]: Addressed, see https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/paul/update-perf-testing-documentation/uiautomationutilities/perf_tool/readme.md#changes-needed-in-common---javakotlin-code

File: uiautomationutilities/perf_tool/readme.md (hunk 12 above)
Comment (paulkagiri) [reply uiautomationutilities/perf_tool/readme.md]: Addressed

File: uiautomationutilities/perf_tool/readme.md (hunk 13 above)
Comment (paulkagiri) [reply uiautomationutilities/perf_tool/readme.md]: Addressed

File: uiautomationutilities/perf_tool/readme.md (hunk 16 above)
Comment (paulkagiri) [reply uiautomationutilities/perf_tool/readme.md]: Done

File: uiautomationutilities/perf_tool/readme.md (hunk 17 above)
Comment (paulkagiri) [reply uiautomationutilities/perf_tool/readme.md]: I believe you can also run Visual Studio on Mac

File: uiautomationutilities/perf_tool/readme.md (hunk 15 above)
Comment (paulkagiri) [reply uiautomationutilities/perf_tool/readme.md]: Ordered ok now

**Review by shoatman — APPROVED** [review]

File: uiautomationutilities/perf_tool/readme.md [hunk 18]
```diff
@@ -1,155 +1,164 @@
...
-- 3. Build id of the base task.
-- 4. Build id which should be written in the final Email report and used for going to artifact url. Example value: "1234"
-- 5. Device model to be written in the final Email report. Example value: "Pixel2"
-- 6. Device OS to be written in the final Email report. Example value: "API28"
-- 7. App name to be written in the Email report. Example value: "MSALTestApp"
-- 8. Email ID of the sender's account. Example value: "idlab1@msidlab4.onmicrosoft.com"
-- 9. Password of the sender's account.
--10. Email To list separated by comma
+- Directory where PerfData base files are present. Example value: `C:\testdata\basefiles`
+- Directory where PerfData target files are present. Example value: `C:\testdata\targetfiles`
```
Comment (AdamBJohnsonx) [inline uiautomationutilities/perf_tool/readme.md:153]: Does this run on non-windows machines?

Comment (paulkagiri) [reply uiautomationutilities/perf_tool/readme.md:153]: With some modifications it could but it has only been tested on Windows - https://github.com/AzureAD/microsoft-authentication-library-common-for-android/blob/paul/update-perf-testing-documentation/uiautomationutilities/perf_tool/readme.md#steps-to-deploy-latest-version-of-c-tool-to-pipeline

File: uiautomationutilities/perf_tool/readme.md (hunk 2 above)
Comment (AdamBJohnsonx) [reply uiautomationutilities/perf_tool/readme.md:11]: We're migrating our pipelines there.  We should look at doing that or making sure there's a task to do so.

File: uiautomationutilities/perf_tool/readme.md (hunk 11 above)
Comment (AdamBJohnsonx) [reply uiautomationutilities/perf_tool/readme.md:55]: I would be curious to know the depth of the any change required to abandon the numeric backing of these strings.  Because this seems like a completely unnecessary extra step.  Not that this matters for these instructions, but that's where we are.

**Review by AdamBJohnsonx — APPROVED** [review]

File: uiautomationutilities/perf_tool/readme.md (hunk 2 above)
Comment (paulkagiri) [reply uiautomationutilities/perf_tool/readme.md:11]: Will create a task to do so
